common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,206 @@
+     * @param loopContinuously  whether to keep the thread look for file modification after one iteration. This option
+     *                          is useful for testing only.
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws NullPointerException if either {@code fileToWatch}  or {@code callback} is null
+     */
+    @VisibleForTesting
+    FileModificationEventWatcher(Path fileToWatch, Consumer<WatchEvent<?>> callback,
+                                 boolean loopContinuously)
+            throws FileNotFoundException {
+        super();
+        this.validateInput(fileToWatch, callback);
+
+        this.pathOfFileToWatch = fileToWatch;
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+
+        // Set the name for this object/thread for identification purposes.
+        this.setName("file-update-watcher-" + THREAD_NUM.incrementAndGet());
+        this.loopContinuously = loopContinuously;
+    }
+
+    @VisibleForTesting
+    String nameOfFileToWatch() {
+        Path fileName = this.pathOfFileToWatch.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            throw new RuntimeException("File name is null");
+        }
+    }
+
+    @VisibleForTesting
+    Path directoryOfFileToWatch() {
+        assert this.pathOfFileToWatch != null;
+        return this.pathOfFileToWatch.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.pathOfFileToWatch);
+
+            String fileName = nameOfFileToWatch();
+            Path directoryPath = directoryOfFileToWatch();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.pathOfFileToWatch);
+
+            while (true) {
+                watchKey = watchService.take();
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.pathOfFileToWatch);
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
+                Thread.sleep(200);
```

Comment:
Each file modification/create usually results in the WatcherService reporting the WatchEvent twice, as the file is updated twice: once for the content and once for the file modification time. These events occur in quick succession. We wait for 200 ms. here so that the events get
de-duplicated - in other words only single event is processed.

If the two events are not raised within this duration, the callback will be invoked twice, which we assume is not a problem for applications of this object. In its current use for reloading the SSL Context, that would mean reloading the context twice, which doesn't break anything; It's just that the SSL context is reloaded twice unnecessarily - reloading it once is sufficient.